phantom_token.c

```diff
@@ -628,54 +628,68 @@ static ngx_int_t introspection_response_handler(ngx_http_request_t *request, voi
         return introspection_subrequest_status_code;
     }
 
-    // body parsing
     u_char *jwt_start = NULL;
+    ngx_str_t cache_data = ngx_null_string;
 
+#if (NGX_HTTP_CACHE)
     if (!request->cache || !request->cache->buf)
     {
+        // No cache; read JWT from response to sub-request
         jwt_start = request->header_end + sizeof("\r\n") - 1;
     }
 
     if (jwt_start == NULL && request->cache && request->cache->buf && request->cache->valid_sec > 0)
     {
-        ngx_read_file(&request->cache->file, request->cache->buf->pos, request->cache->length, 0);
+        // Try to read JWT from cache
 
-        jwt_start = request->cache->buf->start + request->cache->body_start;
+        cache_data.len = request->cache->length;
+        cache_data.data = ngx_pnalloc(request->pool, cache_data.len);
+
+        if (cache_data.data != NULL)
+        {
+            ngx_read_file(&request->cache->file, cache_data.data, cache_data.len, request->cache->body_start);
```

Comment:
This is the fix. `request->cache->buf` was being freed by NGINX after the first request. When we touched that memory on the second request, we'd crash.